mon,osd,mds: add "{mon,mds} metadata" and "node ls" commands

tchaikov opened the pull request:
the original intent is to add "node ls" command. (see http://tracker.ceph.com/issues/10904)

this PR was #4100 .

jcsp commented on src/mon/MDSMonitor.cc:
```diff
@@ -1680,6 +1694,80 @@ void MDSMonitor::check_sub(Subscription *sub)
+    return r;
+  }
+  if (!metadata.count(gid)) {
+    return -EINVAL;
```
Probably should be -ENOENT

jcsp replied:
Should populate err with message on error

jcsp commented on src/mon/MonCommands.h:
```diff
@@ -235,6 +235,10 @@ COMMAND("report name=tags,type=CephString,n=N,req=false", \
 	"mon", "r", "cli,rest")
 COMMAND("quorum_status", "report status of monitor quorum", \
 	"mon", "r", "cli,rest")
+COMMAND("mon_metadata name=rank,type=CephInt,range=0",
```
inconsistent use of underscores vs spaces in commands (including pre-existing commands) -- I'd suggest all the new commands should use spaces.

tchaikov replied:
thank you @jcsp . yes, that's actually what i came up first. but i changed it to "mon_metadata" later on. 

because, "$foo $bar" command are handled by module named `$foo` where $foo can be mds,fs,osd,pg,mon,auth ... . in the case of mon, the `$bar` command will be handled by `MonmapMonitor` which is a PaxosService, but the implementation of `mon_metadata` lives in `Monitor` itself. and a plain Dispatcher will serve my implementation well as suggested by @jecluis . maybe i can rework the way how Monitor dispatches commands so we are not limited by this constraint?

what do you think?

jcsp commented on src/mon/MDSMonitor.cc:
```diff
@@ -767,6 +768,19 @@ bool MDSMonitor::preprocess_command(MMonCommand *m)
+    }
+    if (!f)
+      f.reset(Formatter::create("json-pretty"));
+    f->open_object_section("mon_metadata");
```
s/mon_metadata/mds_metadata/

jcsp commented on src/mon/MDSMonitor.cc:
```diff
@@ -767,6 +768,19 @@ bool MDSMonitor::preprocess_command(MMonCommand *m)
+    if (!f)
+      f.reset(Formatter::create("json-pretty"));
+    f->open_object_section("mon_metadata");
+    r = get_mds_metadata(gid, f.get(), ds);
```
I think you want to pass 'ss' instead of 'ds' as the "err" argument here

jcsp replied:
we mostly use "get" for methods that return data structures, and "dump" for methods that write out to a Formatter, so this should be dump_mds_metadata

jcsp replied:
...or just "dump_metadata" as the "mds" part is implicit since we're in the scope of MDSMonitor

jcsp commented:
There is nothing removing the MDS metadata for GIDs that no longer exist (see the various cases in which we do an mds_info.erase(...))

jcsp commented on src/mon/MDSMonitor.cc:
```diff
@@ -1680,6 +1694,80 @@ void MDSMonitor::check_sub(Subscription *sub)
+{
+  assert(f);
+
+  f->open_object_section("mds");
```
You're outputting an object with N keys, where you should be outputting an array of N objects with one key each, or an array of N strings.  For example, on my system (with many services on the same hostname):
"""
    {
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain"
    },
"""

That's actually unreadable JSON because when we load it all the "hostname" keys collapse into one.

jcsp commented:
The "node ls" output on a vstart cluster is missing mons for me, and printing "Unable to load metadata":
```
./ceph node ls all
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
Unable to load metadata.
[
    {
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain"
    },
    {
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain"
    }
]
```

(I just happen to know that those two lists are OSDs and MDSs based on the number of items in the list)

jcsp commented on src/mon/MDSMonitor.h:
```diff
@@ -130,11 +130,18 @@ class MDSMonitor : public PaxosService {
+  void update_mds_metadata(mds_gid_t gid, const Metadata& metadata);
+  int load_metadata(map<mds_gid_t, Metadata>& m);
+
+private:
```
redundant "private:"

jcsp commented on src/mon/OSDMonitor.cc:
```diff
@@ -989,11 +988,37 @@ int OSDMonitor::dump_osd_metadata(int osd, Formatter *f, ostream *err)
+void OSDMonitor::print_nodes(Formatter *f)
+{
+  f->open_object_section("osd");
+  for (int osd = 0; osd <= osdmap.get_max_osd(); osd++) {
```
It is usual that a host contains multiple OSDs (and that's possible for mons and MDSs too, although less common).  if you're trying to print a list of hostnames, then you should avoid printing the same one twice.  OTOH it might be more useful to list the individual services here, and their hostnames, like:

```
{
  "host1": ["osd.0", "osd.1"]
  "host2": ["osd.2", "osd.3"]
}
```

tchaikov replied:
the group-by-host feature sounds great. but in a large cluster, to group the OSD hosts could be expensive, i guess.

jcsp commented on src/mon/MDSMonitor.cc:
```diff
@@ -767,6 +768,19 @@ bool MDSMonitor::preprocess_command(MMonCommand *m)
     }
+  } else if (prefix == "mds metadata") {
+    mds_gid_t gid = MDS_GID_NONE;
+    if (!cmd_getval(g_ceph_context, cmdmap, "gid", gid)) {
```
Users rarely want to refer to their daemons by GID.  Ideally, this would do the same trick as MDSMonitor::fail_mds, to take a parameter that can be an ID, GID or rank.

tchaikov commented:
@jcsp thanks for your review. i addressed all of your comments except:

- inconsistent use of underscores vs spaces in commands: the constraint of the way Monitor dispatches commands
-  if you're trying to print a list of hostnames, then you should avoid printing the same one twice: could be expensive for large cluster.
- remove an mds in the output after it is removed. i remove it when it is `mds rm`'ed or it is `mds fail`ed. it follows the way how `osd metadata` removes the OSD metadata entries in the monitor storage. because i want to include the MDS that are down as `ods metada` does..

jcsp commented:
> remove an mds in the output after it is removed. i remove it when it is mds rm'ed or it is mds failed. it follows the way how osd metadata removes the OSD metadata entries in the monitor storage. because i want to include the MDS that are down as ods metada does..

Unfortunately that doesn't quite work in the MDS case, because you're storing these by GID, and the MDS daemon GID changes whenever it restarts.  So with the current code, you get an ever-growing list of MDSs in the "node ls" output.

I think it's okay to only have metadata for MDS daemons that are present in the mds_info map.

jcsp commented:
Just re-tested, this still fails to output mons in "node ls" and prints "Unable to load metadata." instead

jcsp commented:
> if you're trying to print a list of hostnames, then you should avoid printing the same one twice: could be expensive for large cluster.

I don't think this is necessarily expensive -- you can build this in a single pass if you construct a map of (hostname)->(list of services) and iterate through the services once to populate it.

Currently, if I have three OSD hosts with two OSDs each, the output I get is like this:
```
[
 "host1",
 "host1",
 "host2",
 "host2",
 "host3",
 "host3"
]
```

...which just doesn't look very sensible to me.

IMHO there are two sane options for doing this.  You can either list all the services by ID (osd.0, osd.1 etc) and their hostname (which might appear more than once), or you can list all the hostnames (uniquely) with a list of which services appear on that host.  So, calling those options A and B, and using yaml-esque formatting to be quicker it would be:

```
A

osd.0: host1
osd.1: host1
osd.2: host2
osd.3: host2
osd.4: host3
osd.5: host3

*or*

B
host1: osd.0, osd1
host2: osd.2, osd.3
host3: osd.4, osd.5
```

My personal preference would be for option B.

(Other peoples opinions welcome too!)

tchaikov commented:
@jcsp john, i addressed all your comments.
- the output format of hosts and service name, i choose option B. yes, a map with thousands of OSD should be fine. and it's a temporary object.
- the missing-mons issue in `node ls` should be fixed in the latest patch.

thanks for your review and tests. =)

tchaikov commented:
wait! the mon_metadata command is broken.

tchaikov commented:
thanks for your review and suggestions @jecluis , i addressed all your comments in the latest PR.

tchaikov commented:
morning @jecluis ! thanks and sorry, i was using this branch as a stash place to move my changes from my dev box to a faster machine for testing and compilation but bloody slow for editing. i am using another branch for testing now.

dachary commented:
@tchaikov trimmed the network related bot false negative

tchaikov commented:
@jecluis , thanks. doh! i forgot to call i-forgot-the-name-of-the-method. =D

and all your comments are addressed in the latest PR.

jecluis commented:
@tchaikov changes to MMonMetadata look weird.

You initially add MMonMetadata in 6405aeb8ab7c3d9cda7622917d03b96dfb9dc20e, and then change MMonMetada in bf52f5004219439858923f4200040435e1904885, only to change it again in 3aabb2fb5ad2b521480ed3dfb964fe5ec2551a4f.

All those changes to MMonMetadata can be squashed into one single commit (likely the first commit in 6405aeb8ab7c3d9cda7622917d03b96dfb9dc20e).  This means that you will have to extract those changes from the individual patches and squash them.

Aside from that, it looks good.

tchaikov commented:
@jecluis thanks for your time. i cleaned up the commits. and fixed the message leak in https://github.com/tchaikov/ceph/commit/6ed048e1268aaa88ce3f0e71a3e52ff806051308#diff-d4e1ba36dd08617ea271ab2be3dbcb5bR4208.

tchaikov commented:
@jecluis would you mind take a look again? thanks.

jecluis commented:
just the two comments above;  I don't see anything obviously wrong with it, but on further thought it would be really nice if you could make this command based on 'id' instead of 'rank'.

tchaikov commented:
thank you @jecluis ! i updated this pull request to use make `mon_metadata` to use "id" (a,b,c, ...) instead of "rank" (1,2,3,...).

jecluis commented:
@tchaikov looks good!
